cpp/src/arrow/util/value_parsing.h

```diff
@@ -329,13 +373,31 @@ struct StringToSignedIntConverterMixin {
     if (ARROW_PREDICT_FALSE(length == 0)) {
       return false;
     }
+    
+    // If its starts with 0x then its hex
+    if (*s == '0' && ((*(s + 1) == 'x') || (*(s + 1) == 'X'))){
+      length -= 2;
+      s += 2;
+
+      // lets make sure that the length of the string is not too big
+      if (!ARROW_PREDICT_TRUE(sizeof(unsigned_value)*2 >= length && length > 0)) {
```

Comment:
This could be part of `ParseHex` instead of being repeated twice.

Reply:
Done